### PR TITLE
caffe2: fix PinnedCPUAllocator cudaHostRegister() leak

### PR DESCRIPTION
**Pull request by hartb**
In the NUMA case, PinnedCPUAllocator's allocate() would return a
DataPtr constructed by DefaultCPUAllocator, which would reference
the Default... Delete() rather than the Pinned... Delete(). That
meant Pinned... Delete() would never run, so cudaHostUnregister()
would never be called when regions were freed.

See: https://github.com/pytorch/pytorch/issues/16280

This change adds a 'naked_allocate()' method to the Default allocator
that just returns a pointer to the allocated memory rather than
wrapping it in a DataPtr. Pinned allocator uses that then constructs
a DataPtr with reference to its own Delete().

### CODE REVIEW

**Comment by ezyang** [comment]
Thanks a lot, and nice catch.

Just fresh in master, there is a much more compact way to do an equivalent thing: on DataPtr use the method `compare_exchange_deleter`, passing in the old deleter (for the plain CPU allocator) and the new pinned deleter. You might have to make the old deleter public if it isn't already. Would you mind trying that out? Otherwise I will make an attempt.

**Comment by ezyang** [comment]
cc @jerryzh168

**Comment by hartb** [comment]
Thank you for having a look. Sure; we'll try a `compare_exchange_deleter` version.

**Comment by jerryzh168** [comment]
Nice catch! Thanks! @hartb

File: caffe2/core/context_gpu.h [hunk 1]
```diff
@@ -357,14 +357,13 @@ struct CAFFE2_CUDA_API PinnedCPUAllocator final : public at::Allocator {
     at::DataPtr data_ptr;
     std::lock_guard<std::mutex> lock(CUDAContext::mutex());
     if (IsNUMAEnabled()) {
-      data_ptr = baseAllocator_.allocate(nbytes);
-      data = data_ptr.get();
+      data = baseAllocator_.naked_allocate(nbytes);
```
Comment (jerryzh168) [inline caffe2/core/context_gpu.h]: Actually, we are in the process of merge pytorch and caffe2 allocator and we'll be using `Allocator*` for `baseAllocator_`, see: https://github.com/pytorch/pytorch/pull/14517/files#diff-6286b32ea83ee15c66db129928f27c42R343

**Comment by ezyang** [comment]
Hmm, yes, you're right. I guess you have two options:

1. Union Delete and ReportAndDelete into one function, which checks `FLAGS_caffe2_report_cpu_memory_usage` to see if it should report or not
2. Do a condition on `FLAGS_caffe2_report_cpu_memory_usage` before doing a `compare_exchange_deleter` so that you know which deleter to test for in each case.

`FLAGS_caffe2_report_cpu_memory_usage` is a big stick in my craw and I want to get rid of it if possible.

**Comment by ezyang** [comment]
@hartb Are you planning to look at this, or should one of us adopt this patch? Thanks!

**Comment by hartb** [comment]
I hope to update the PR with the suggestion above.

I'm thinking maybe use `baseAllocator_`'s `raw_deleter()` to get the expected Delete function, `CAFFE_ENFORCE()` that the swap succeeds, and include a comment that if the _does_ succeed in the `Report...` case, then we leak a Reporter allocation (and in any case probably break reporting), but that's preferrable to leaking `cudaHostRegister()`.

It's taking a bit longer than I'd hoped to swap my build setup from 1.0.0 to master; I'll let you know if that derails me for some reason.

**Comment by hartb** [comment]
@ezyang Here's a proposed fix (one commit) implementing the above based on PR 14517 (as @jerryzh168 ) mentioned above:

https://github.com/hartb/pytorch/tree/hartb-pr14517-add

Would you like to pick this over to that PR, or should modify that to base on master via this PR?

Note my worry above about leaking or breaking the Reporting case isn't an issue. The Pinned Delete() calls the base Allocator's delete to finish the job, so we get base/Default allocator clean-up that way.

**Comment by ezyang** [comment]
@hartb Feel free to force push! Sorry about the delay responding.

**Comment by hartb** [comment]
Will update this PR once tested the fix rebased to master.

**Review by ezyang — APPROVED** [review]
Sweet and simple

**Review by dzhulgakov — APPROVED** [review]

**Pull request closed without merging**